### PR TITLE
Added doc-comments to tensor.rs

### PR DESCRIPTION
**Pull request by nestordemeure**
Following the discussion in [issue11](https://github.com/elodin-sys/elodin/issues/11), here is a PR adding doc comments to `tensor.rs`. This is a first test in introducing AI-generated comments to document the library faster, making it easier to pick up by new users.

Here is a preview of what it looks like in conditions (`cargo doc --open --no-deps --all-features`):

image
image

This is a test to think about the viability of this approach. Do feel free to comment and criticize (things can definitely be streamlined by hand, this PR had no manual tweak of the comments so far)!

For archival purposes, here is the prompt used to generate the comments (with GPT-4):

```
You will be given code from `nox`, a Rust implementation of JAX. Your goal is to doc-comment the code:
* adding a top-level doc comment (describing the current file's function),
* doc comment the functions (helping users understand the arguments and outputs of the function),
* doc comment types and traits.

Your output should be only the doc comments and headers, no need to include the rest of the code.

Also, keep your comments short and useful. No point in stating the obvious or writing a long sentence where a short one would do. Those comments should be there to help a new user understand how to use the library from a quick look.
```

### CODE REVIEW

File: libs/nox/src/tensor.rs [hunk 1]
```diff
@@ -389,10 +406,16 @@ impl<T, D: TensorDim> AsBuffer for Tensor<T, D, Buffer> {
     }
 }
 
+/// Trait for mapping dimensions in tensor operations.
+/// Allows for transforming and replacing dimensions in tensor types.
 pub trait MapDim<D> {
+    /// The item associated with the dimension mapping.
```
Comment (akhilles) [inline libs/nox/src/tensor.rs]: We probably don't need these inner comments.

File: libs/nox/src/tensor.rs [hunk 2]
```diff
@@ -495,7 +520,9 @@ impl<const N: usize> DefaultMap for Const<N> {
     type DefaultMapDim = Mapped;
 }
 
+/// Trait representing the concatenation of dimensions.
 pub trait DimConcat<A, B> {
+    /// The resulting dimension type after concatenation.
```
Comment (akhilles) [inline libs/nox/src/tensor.rs]: This is implied.

File: libs/nox/src/tensor.rs [hunk 3]
```diff
@@ -640,9 +671,12 @@ impl<T: TensorItem, D: TensorDim> Tensor<T, D> {
     }
 }
 
+/// Trait for indexing into tensors, allowing for the extraction of sub-tensors or elements based on indices.
 pub trait TensorIndex<T, D: TensorDim> {
+    /// The output type resulting from the indexing operation.
```
Comment (akhilles) [inline libs/nox/src/tensor.rs]: This is implied.

**Review by akhilles — COMMENTED** [review]
This is pretty good! The top level comments are all fine. Some of the comments on associated types and trait functions seem a bit unnecessary. Especially for `Out`, `Output`. @sphw, could you take a look as well?

**Comment by akhilles** [comment]
The commit message has a typo btw

**Comment by nestordemeure** [comment]
Thanks, I have fixed the commit text and removed inner comments.

File: libs/nox/src/tensor.rs [hunk 4]
```diff
@@ -10,6 +11,8 @@ use std::{
 };
 use xla::{ArrayElement, ElementType, NativeType};
 
+/// Represents a tensor with a specific type `T`, dimensionality `D`, and parameterization `P`.
```
Comment (akhilles) [inline libs/nox/src/tensor.rs]: "underlying representation `P`" is a bit more accurate.

File: libs/nox/src/tensor.rs [hunk 5]
```diff
@@ -10,6 +11,8 @@ use std::{
...
+/// `P` dictates the underlying representation and operations available on the tensor.
```
Comment (akhilles) [inline libs/nox/src/tensor.rs]: "operations available on the tensor" is not quite accurate. Would just drop this sentence, and change the previous sentence to include the "underlying representation" part.

**Review by akhilles — COMMENTED** [review]
Just have a nitpick on the `Tensor` comment, but ok to merge once that's fixed.

File: libs/nox/src/tensor.rs (hunk 5 above)
Comment (nestordemeure) [reply libs/nox/src/tensor.rs]: Done!

**Review by akhilles — APPROVED** [review]

**Pull request merged**